src/main/scala/cromwell/engine/workflow/ValidateActor.scala

```diff
@@ -38,29 +49,22 @@ class ValidateActor(wdlSource: WdlSource, workflowInputs: Option[WdlJson], workf
   override def receive = {
     case ValidateWorkflow =>
       validateWorkflow(sender())
-      // NOTE: self shuts down when the parent PerRequest shuts down
-  }
-
-  private def validateInputs(namespaceWithWorkflow: NamespaceWithWorkflow, maybeWorkflowInputs: Option[WdlJson]): Future[Unit] = {
-    maybeWorkflowInputs match {
-      case Some(wi) =>
-        for {
-          inputs <- Future(wi.parseJson).map(_.asJsObject.fields)
-          coercedInputs <- Future.fromTry(namespaceWithWorkflow.coerceRawInputs(inputs))
-        } yield ()
-      case None => Future.successful(())
-    }
+    // NOTE: self shuts down when the parent PerRequest shuts down
   }
 
   private def validateWorkflow(sentBy: ActorRef): Unit = {
     logger.info(s"$tag for $sentBy")
-    val futureValidation: Future[Unit] = for {
+    val futureValidation: Future[ValidationResult] = (for {
       namespaceWithWorkflow <- Future(NamespaceWithWorkflow.load(wdlSource))
-      validatedInputs <- validateInputs(namespaceWithWorkflow, workflowInputs)
-      validatedRuntimeOptions <- Future.fromTry(TryUtil.sequence(namespaceWithWorkflow.workflow.calls map {
-        call => CromwellRuntimeAttributes.validateKeys(call.task.runtimeAttributes.attrs.keySet, CromwellBackend.backend.backendType)
-      }))
-    } yield ()
+      validatedInputs: Option[WorkflowCoercedInputs] <- Future(validateInputs(namespaceWithWorkflow))
+      validatedWorkflowOptions: Option[WorkflowOptions] <- Future(validateWorkflowOptions(CromwellBackend.backend()))
+      validatedRuntimeAttrs: Seq[Set[String]] <- Future(validateRuntimeAttributes(namespaceWithWorkflow))
+    } yield ValidationSuccess(namespaceWithWorkflow, validatedInputs, validatedWorkflowOptions, validatedRuntimeAttrs)) recover {
+      case reason: Throwable => ValidationFailure(reason)
+    }
+
+    //TODO: IMO, the responsibility for returning a status code should be a part of the client. This class should only return a validation result,
+    // and the callee should convert that to a status code or anything that it wants
 
     // Now validate that this Future completed successfully:
     futureValidation onComplete {
```

Comment:
Now the `Future` can't fail because of the recover, so I think this should be something like

```scala
    // Now validate that this Future completed successfully:
    futureValidation map {
      case _: ValidationSuccess =>
        logger.info(s"$tag success $sentBy")
        sentBy ! RequestComplete(
          StatusCodes.OK,
          APIResponse.success("Validation succeeded."))

      case ValidationFailure(ex) =>
        val messageOrBlank = Option(ex.getMessage).mkString
        logger.info(s"$tag error $sentBy: $messageOrBlank")
        sentBy ! RequestComplete(
          StatusCodes.BadRequest,
          APIResponse.fail(ex))
    }
```